src/quadiron-c.cpp

```diff
@@ -0,0 +1,450 @@
+    int block_size)
+{
+    quadiron::fec::RsFnt<uint32_t>* fec =
+        reinterpret_cast<quadiron::fec::RsFnt<uint32_t>*>(fecp);
+    std::vector<uint8_t*> data_vec(fec->n_data, nullptr);
+    std::vector<uint8_t*> parities_vec(fec->n_outputs, nullptr);
+    std::vector<quadiron::Properties> parities_props(fec->n_outputs);
+    std::vector<int> missing_idxs_vec(
+        missing_idxs, missing_idxs + fec->code_len);
+    std::vector<int> wanted_idxs_vec(fec->n_data, 1);
+    int metadata_size = quadiron_fnt32_get_metadata_size(fecp, block_size);
+    bool res;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                parities_vec[i] = parity[i] + metadata_size;
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret =
+                    parities_props[i].from_buffer(metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+        }
+    } else {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (!missing_idxs[i]) {
+                parities_vec[i] = data[i] + metadata_size;
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(data[i]);
+                int ret =
+                    parities_props[i].from_buffer(metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].from_buffer(
+                    metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+        }
+    }
+
+    res = fec->decode_blocks_vertical(
+        data_vec,
+        parities_vec,
+        parities_props,
+        missing_idxs_vec,
+        wanted_idxs_vec,
+        block_size);
+    if (!res)
+        return -1;
+
+    return 0;
+}
+
+/**
+ * Reconstruct block
+ *
+ * @param data must be exactly n_data (use nullptr when missing) -
+ * buffers must allocate block_size + metadata_size
+ * @param parity must be exactly n_parities - buffers must allocate block_size +
+ * metadata_size
+ * @param missing_idxs array of missing_idxs of len code_len indicating presence
+ * (value 1) or absence of fragments (value 0)
+ * @param destination_idx index of fragment to reconstruct
+ * @param block_size the block size in bytes
+ *
+ * @return 0 if reconstruct succeeded, else -1
+ */
+int quadiron_fnt32_reconstruct(
```

Comment:
this function si too big, it should be split.

Maybe like that:
- prologue (preparing the variables, maybe factored out with `decode` (see my other comment))
- no rebuild case
- rebuild case
- prologue (finalizing the variables, maybe factored out with `encode` (see my other comment))

Or like that:
- prologue
- systematic
- non-systematic
- epilogue

Or in another better way.